rust/actyx/ax-aql/src/language/types.rs

```diff
@@ -0,0 +1,108 @@
+use super::{non_empty::NonEmptyString, parse_utils::P, parser::Rule};
+use crate::{
+    language::{
+        parse_utils::{Ext, Spanned},
+        parser::{r_bool, r_nonempty_string, r_string, NoVal},
+    },
+    NonEmptyVec,
+};
+use once_cell::sync::Lazy;
+use pest::pratt_parser::{Assoc, Op, PrattParser};
+use std::sync::Arc;
+
+#[derive(Debug, PartialEq, Eq, PartialOrd, Ord, Clone)]
+pub enum Type {
+    Atom(TypeAtom),
+    Union(Arc<(Type, Type)>),
+    Intersection(Arc<(Type, Type)>),
+    Array(Arc<Type>),
+    Dict(Arc<Type>),
+}
+
+#[derive(Debug, PartialEq, Eq, PartialOrd, Ord, Clone)]
+pub enum TypeAtom {
+    Null,
+    Bool(Option<bool>),
+    Number(Option<u64>),
```

Comment:
Doesn't this make all numbers, necessarily a `u64`?

Reply:
Adding support for proper numbers is still possible, but I defer it because it is a huge effort.
BigInt/Decimal/Float are already supported by cbor-data, but all computation is still missing.

Reply:
So, refinements only work on u64 and the other cases can be floats?

Reply:
No, currently the grammar and parser only accepts u64, which is a proper subset, so it can be expanded in the future.